lib/IDE/IDETypeChecking.cpp

```diff
@@ -731,6 +731,78 @@ swift::collectExpressionType(SourceFile &SF,
   return Scratch;
 }
 
+/// This walker will traverse the AST and report types for every variable declaration.
+class VariableTypeCollector: public SourceEntityWalker {
+private:
+  SourceManager &SM;
+  unsigned int BufferId;
+  std::vector<VariableTypeInfo> &Results;
+
+  /// This is where we print all types.
```

Comment:
Since the entire handling of type strings is a little bit confusing at first, I think it would be good to add some background documentation here, like.

```cpp
In a SourceKit request, we transfer all type names in a single string. To describe individual types, we use offsets into that big string. `OS` builds a string that contains all null-terminated type strings. When referring to one of these types, we can use the offset at which it starts in `OS`.
```

Reply:
I thought `VariableTypeArray` used the buffer + offset to transfer the types or am I mistaken?

Reply:
In `VariableTypeReader` ([`VariableTypeArray.cpp`](https://github.com/fwcd/swift/blob/43b6c0a0a27fbe9a087608d037c73ef1060777a8/tools/SourceKit/tools/sourcekitd/lib/API/VariableTypeArray.cpp#L87)), the sourcekitd dictionaries are built by extracting the types at their corresponding offsets from the buffer.

Reply:
Ah, I see. In that case I *think* that we are using the output stream to avoid allocating new memory for each string. I think @nkcsgexi implemented `ExpressionTypeInfo`, maybe he’s got more background. 

In any case, I think it would be cleaner to use `StringRef`s instead of the offsets we are using at the moment.